compiler/x/amd64/objectfmt/OMRJitCodeRWXObjectFormat.cpp

```diff
@@ -0,0 +1,217 @@
+          * The only situation where targetAddress could be 0 at this point is for
+          * a recursive call.  This is confirmed/protected by the fatal assert above.
+          * Emit a direct call to the start of the function whose displacement
+          * will be properly relocated by the binary encoding of the instruction.
+          */
+         if (data.prevInstr)
+            {
+            callInstr = generateImmSymInstruction(data.prevInstr, CALLImm4, 0, data.methodSymRef, data.regDeps, data.cg);
+            }
+         else
+            {
+            callInstr = generateImmSymInstruction(CALLImm4, data.callNode, 0, data.methodSymRef, data.regDeps, data.cg);
+            }
+         }
+      else
+         {
+         /**
+          * Ideally, only the RegImm64Sym instruction form would be used.  However, there are
+          * subtleties in the handling of relocations between these different forms of instructions
+          * that need to be sorted out first..
+          */
+         if (data.useSymInstruction)
+            {
+            TR::AMD64RegImm64SymInstruction *loadInstr;
+
+            if (data.prevInstr)
+               {
+               loadInstr = generateRegImm64SymInstruction(
+                  data.prevInstr,
+                  MOV8RegImm64,
+                  data.scratchReg,
+                  targetAddress,
+                  data.methodSymRef,
+                  data.cg);
+               }
+            else
+               {
+               loadInstr = generateRegImm64SymInstruction(
+                  MOV8RegImm64,
+                  data.callNode,
+                  data.scratchReg,
+                  targetAddress,
+                  data.methodSymRef,
+                  data.cg);
+               }
+
+            if (data.reloKind != TR_NoRelocation)
+               {
+               loadInstr->setReloKind(data.reloKind);
+               }
+
+            data.out_loadInstr = loadInstr;
+            }
+         else
+            {
+            TR::AMD64RegImm64Instruction *loadInstr;
+
+            if (data.prevInstr)
+               {
+               loadInstr = generateRegImm64Instruction(
+                  data.prevInstr,
+                  MOV8RegImm64,
+                  data.scratchReg,
+                  targetAddress,
+                  data.cg,
+                  data.reloKind);
+               }
+            else
+               {
+               loadInstr = generateRegImm64Instruction(
+                  MOV8RegImm64,
+                  data.callNode,
+                  data.scratchReg,
+                  targetAddress,
+                  data.cg,
+                  data.reloKind);
+               }
+
+            data.out_loadInstr = loadInstr;
+            }
+
+         TR_X86OpCodes op = data.isCall ? CALLReg : JMPReg;
+         if (data.prevInstr)
+            {
+            callInstr = generateRegInstruction(data.out_loadInstr, op, data.scratchReg, data.regDeps, data.cg);
+            }
```

Comment:
Perhaps I'm not familiar enough with this area of the code, but this looks a bit unintuitive to me. Every other occurrence of this pattern passes `data.pervInstr` to `generate*Instruction`, but here we pass `data.out_loadInstr`. Maybe a comment explaining the reason would help?